tamr_client/operation.py

```diff
@@ -0,0 +1,195 @@
+
+    Synchronous mode:
+        Automatically waits for operation to resolve before returning the
+        operation.
+
+    asynchronous mode:
+        Immediately return the ``'PENDING'`` operation. It is
+        up to the user to coordinate this operation with their code via
+        :func:`~tamr_client.operation.wait` and/or
+        :func:`~tamr_client.operation.poll` .
+
+    Args:
+        asynchronous: Whether or not to run in asynchronous mode. Default: ``False``.
+        ``**options``: When running in synchronous mode, these options are
+                passed to the underlying :func:`~tamr_client.operation.wait` call.
+    """
+    if asynchronous:
+        return operation
+    return wait(session, operation, **options)
+
+
+def _from_url(session: Session, url: URL) -> Operation:
+    """Get operation by URL
+
+    Fetches operation from Tamr server
+
+    Args:
+        url: Operation URL
+
+    Raises:
+        OperationNotFound: If no operation could be found at the specified URL.
+            Corresponds to a 404 HTTP error.
+        requests.HTTPError: If any other HTTP error is encountered.
+    """
+    r = session.get(str(url))
+    if r.status_code == 404:
+        raise NotFound(str(url))
+    data = response.successful(r).json()
+    return _from_json(url, data)
+
+
+def _from_json(url: URL, data: JsonDict):
+    """Make operation from JSON data (deserialize)
+
+    Args:
+        url: Operation URL
+        data: Operation JSON data from Tamr server
+    """
+    cp = deepcopy(data)
+    return Operation(
+        url,
+        id=cp["id"],
+        type=cp["type"],
+        status=cp.get("status"),
+        description=cp.get("description"),
+    )
+
+
+def poll(session: Session, operation: Operation) -> Operation:
+    """Poll this operation for server-side updates.
+
+    Does not update the :class:`~tamr_client.operation.Operation` object.
+    Instead, returns a new :class:`~tamr_client.operation.Operation`.
+
+    Args:
+        operation: Operation to be polled.
+    """
+    return _from_url(session, operation.url)
+
+
+def wait(
+    session: Session,
+    operation: Operation,
+    *,
+    poll_interval_seconds: int = 3,
+    timeout_seconds: Optional[int] = None,
+) -> Operation:
+    """Continuously polls for this operation's server-side state.
+
+    Args:
+        operation: Operation to be polled.
+        poll_interval_seconds: Time interval (in seconds) between subsequent polls.
+        timeout_seconds: Time (in seconds) to wait for operation to resolve.
+
+    Raises:
+        TimeoutError: If operation takes longer than `timeout_seconds` to resolve.
+    """
+    started = now()
+    while timeout_seconds is None or now() - started < timeout_seconds:
+        if operation.status and operation.status["state"] in ["PENDING", "RUNNING"]:
+            sleep(poll_interval_seconds)
+        elif not operation.status or operation.status["state"] in [
```

Comment:
we can probably simplify the logic here by using a guard clause for `operation.status` first:

```python
if operation.status is None:
    return operation
elif operation.status["state"] in ["PENDING", ...]:
    ...
elif operation.status["state"] in ["CANCELED", ...]:
    ...
```